### PR TITLE
Site Selection: Redirect to primary domain

### PR DESCRIPTION
**Pull request by creativecoder**
#### Changes proposed in this Pull Request

* When navigating to a route that used a `*.wordpress.com` subdomain, if the site has a primary mapped domain, redirect to the primary domain.
* For example, for site example.blog as the primary domain, when navigating to `/posts/example.wordpress.com`, redirect to `/posts/example.blog` so Calypso loads the correct site context.

#### Testing instructions

On a site with a mapped domain

* Visit `/site-editor/example.wordpress.com` (using the correct wordpress.com subdomain for the site)
* You should end up in the Site editor, rather than on a blank screen.

Related to https://github.com/Automattic/wp-calypso/issues/62915 and #61953

### CODE REVIEW

**Comment by Addison-Stavlo** [comment]
I tried testing this with the original problematic site noted in slack (p1650392267606899-slack-C01DF571R5Y). Ive tried first with the calypso.live link above as well as building this branch locally, and both times they get redirected to the whitescreen at `*/site-editor?site={siteUrl}` 🤔

**Comment by creativecoder** [comment]
> I tried testing this with the original problematic site noted in slack 

Are a user on that site? Unless you are, such that it shows up in your Calypso list of sites, that's expected. It's the fallback behavior for site selection when Calypso can't find the site slug from the url in your list of sites.

**Comment by Addison-Stavlo** [comment]
> Are a user on that site? Unless you are, such that it shows up in your Calypso list of sites, that's expected. 

Ah silly me that makes sense. Testing with my own domain mapped site now I can see this works on my end!

**Review by Addison-Stavlo — APPROVED** [review]

**Comment by creativecoder** [comment]
Thanks for the review @Addison-Stavlo !

This touches some low level logic for routing, so I'd appreciate additional 👀 if others have time to test and make sure this doesn't break anything.

**Comment by vykes-mac** [comment]
I've also tested this and confirm that it works, did some additional smoke testing and nothing seem to be broken.

**Comment by Addison-Stavlo** [comment]
>  I'd appreciate additional 👀 if others have time to test and make sure this doesn't break anything.

I added a couple of the 'suggested reviewers' to hopefully help here.

**Comment by creativecoder** [comment]
> I added a couple of the 'suggested reviewers' to hopefully help here.

Ah, thanks @Addison-Stavlo ! I lost track of this one.

I'm fairly confident these changes are okay, so I'll wait a day or so for additional feedback, but otherwise go ahead and deploy it.

File: client/my-sites/controller.js [hunk 1]
```diff
@@ -476,6 +477,14 @@ export function siteSelection( context, next ) {
 			.catch( () => null )
 			.then( ( site ) => {
 				let freshSiteId = getSiteId( getState(), siteFragment );
+				const siteSlug = getSiteSlug( getState(), site.ID );
```
Comment (sirbrillig) [inline client/my-sites/controller.js:480]: 🤔  just a drive-by comment but the `.catch()` for this Promise happens _before_ the `.then()` so the `null` returned by the catch is passed as `site` to the `then` callback if there's an error. See:

```js
const myTestCall = () => Promise.reject('error');
myTestCall().catch(() => null).then((input) => console.log('input was ' + input)); // outputs "input was null"
```

It looks like that was intentional in the original code but the newly added code here assumes that `site` is defined and will throw errors.
